### PR TITLE
[Foxy] Handle signals within the asyncio loop. (#476)

### PR DESCRIPTION
**Pull request by mjeronimo**
Signed-off-by: Michel Hidalgo <michel@ekumenlabs.com>

Backporting https://github.com/ros2/launch/pull/476 to Foxy. 
Fixes: https://github.com/ros2/launch_ros/issues/223

### CODE REVIEW

**Comment by hidmic** [comment]
If we backport #476, we have to backport #479, #494, #497, and #498 too. Signal handling is tricky.

**Comment by mjeronimo** [comment]
@hidmic  Do you recommend proceeding with those for Foxy? Is it worth it for this issue?

**Comment by hidmic** [comment]
@mjeronimo all those PRs patch #476 in one way or another. #479 deals with an issue in `asyncio` that became a problem when signal management implementation changed in #476. #494, #497, and #498 fix issues in #476 in various ways (mainly for Windows).

**Comment by mjeronimo** [comment]
@jacobperron  What do you think about backporting this set of changes to Foxy? Would like to get your input before proceeding.

**Comment by jacobperron** [comment]
I'm okay with backporting all of the changes pointed out, as long as @hidmic doesn't have any major reservations about them. On the surface of it, it looks like it's worth the effort to improve things in Foxy, but I don't know enough about the signal handling changes to say if we're going to break users code.

Of all of the changes, the only thing I would change for Foxy would be to keep this function (unused) to maintain API compatibility: 

https://github.com/ros2/launch/pull/494/files#diff-5dcef8ac9a56749cd4f33305fea85b4c918383b74c8c1369cce3a22286d84ee9L30

**Comment by hidmic** [comment]
> as long as @hidmic doesn't have any major reservations about them

No particular reservations. It's sensitive code, that's all.

> Of all of the changes, the only thing I would change for Foxy would be to keep this function (unused) to maintain API compatibility:

That function didn't exist before #476, so it doesn't exist in Foxy. On the other hand, this patch removes quite a few functions in [`signal_management.py`](https://github.com/ros2/launch/pull/506/files#diff-5dcef8ac9a56749cd4f33305fea85b4c918383b74c8c1369cce3a22286d84ee9). Off the top of my head, I don't think we can keep them around. Maybe as a no-op?

**Comment by felixdivo** [comment]
I can probably offer to help with testing these changes. It can't be too hard to set it up properly.

**Comment by felixdivo** [comment]
The fix would be important for me / my team. Can I assist you in some way?

Keeping the old API and having it do nothing seems like the only sensible way (the only one with moderate coding effort). One should perhaps also print a warning pointing to this PR or otherwise explaining that the functionality has moved and the function won't do anything any more. That way, these is actually merit in keeping it.

File: launch/launch/utilities/__init__.py [hunk 1]
```diff
@@ -32,10 +29,7 @@
     'create_future',
     'ensure_argument_type',
     'perform_substitutions',
-    'install_signal_handlers',
-    'on_sigint',
-    'on_sigquit',
-    'on_sigterm',
```
Comment (jacobperron) [inline launch/launch/utilities/__init__.py:38]: Removing these APIs will break anyone using them directly.

**Review by jacobperron — COMMENTED** [review]
I overlooked the changes in the PR originally (and only looked at the follow-up fixes).

I'm very hesitant about backporting this to Foxy since it breaks API. I think we should investigate either:

1. Making an API-compatible fix here (or in launch_ros).
2. Keeping the old API, and keep it functional.

I don't like the idea of making the old API a noop (even if it produces a warning). This can cause a subtle break for the user versus a hard break (if the API does not exist). If we are going to produce a warning, we may as well remove the old API since it forces the user to update instead of potentially having this break subtly.

**Comment by hidmic** [comment]
> Keeping the old API, and keep it functional.

Hmm, perhaps we can use the fact that `asyncio` loops are thread-local globals and re-implement this API with a global `AsyncSignalSafeManager` instance. I haven't thought through all the implications though.

**Comment by felixdivo** [comment]
I just want to ping this issue and find out what the roadmap might be. Is there someone willing to tackle this? I of course unterstand, that this is a rather nasty change and personally also do not want to dig into it.

So if the decision is taken that there simply will not be a fix and foxy will be left with the bug ☹️, it would be great to spell it out in this issue. That way, any stakeholder knows what to do. I my case, I will probably have to hack something together in our product until the next ROS2 LTS is released.

**Comment by hidmic** [comment]
I just put together a POC in #511. It needs testing, and Foxy ROS Boss approval.

**Comment by hidmic** [comment]
Full Foxy CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14824/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9574/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12490/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14979/)

**Comment by felixdivo** [comment]
As with #511, this PR should probably be added to the [Foxy Patch Release 5](https://github.com/orgs/ros2/projects/30)

**Comment by hidmic** [comment]
Test failures seem unrelated. Are we good to go @jacobperron @mjeronimo ?

**Comment by mjeronimo** [comment]
Looks good to me. Thanks for taking on this complicated issue.

**Comment by hidmic** [comment]
@jacobperron I'd like your blessing before I merge this.

**Comment by felixdivo** [comment]
Where should I look for a possible release date of the [Foxy Patch Release 5](https://github.com/orgs/ros2/projects/30) that this will be part of?

**Review by hidmic — APPROVED** [review]

**Review by jacobperron — APPROVED** [review]

**Comment by hidmic** [comment]
Alright, moving forward !

**Pull request merged**